implementation/transformer/tsd_net.py

```diff
@@ -101,17 +101,32 @@ def score(self, hidden, encoder_outputs):
         return energy
 
 
+class PositionalEncoder(nn.Module):
```

Comment:
I believe the idea behind `PositionalEnconder` is concatenating each input token with its position.
However, I do not understand the formulas. Why exp, 10000.0, sin, cos...?
@pixelneo, could you please add some documentation?

Also, where is this positional information being added in `SimpleDynamicEncoder`?
I don’t see the same formulas here. Did you move code from `BSpanDecoder.position_encoding_init` into here?
If yes, why the change?

Reply:
> why exp, 10000.0, sin, cos...?

This comes from the original paper (Vaswani et al. 2017). 10000 is arbitrary, other number would work, but people seem to use it.

> where` is the positional information being added in SimpleDynamicEncoder?

here: `embedded = self.positional_encoder(embedded) # add positional encoding`
 It simply adds it (not concatenates - again both would probably work, but they use summation)

> Did you move code from BSpanDecoder.position_encoding_init into the
    Encoder? If yes, why the change?

Not sure if I understand. I commented out the positional encoding in decoder - not sure if we need it, since there has to be a positional encoding in transfomer anyway.

Reply:
Thank you for the explanation. Very helpful. 👍 

Let me re-state one more time, though: I was interested in the connection between your newly written encoder and the `SimpleDynamicEncoder` from the original Sequicity implementation.

> there has to be a positional encoding in transfomer anyway.

might be the answer I am looking for.

Is it true that you decided to add a positional encoding in the Encoder code even though positional encoding was not present in the Sequicity `SimpleDynamicEncoder`? If I understand correctly, the "positional encoding code" was previously present in `BSpanDecoder`. This raises the question, if it is ok to have position information in `ResponseDecoder`. (I assume it wasn't there, since you didn't have to remove it. Also, my understanding is that both decoders use outputs from the encoder.).

Also, in the new code, you call `positional_encoder` before `transformer_encoder`. Thus, the position information is not directly present in the encoder output. Is this what the original paper does?

Reply:
> Is it true that you decided to add a positional encoding in the Encoder code even though positional encoding was not present in the Sequicity SimpleDynamicEncoder?

Definitely. Because Sequicity uses RNNs, it has the positional information about tokens.
Transformer without positional encoding would be just a bag of words.

About the decoders... i don't know whether we need to add another positional encoding in the bspandecoder, but we definitely need the one from encoder

> you call positional_encoder before transformer_encoder . Thus, the position information is not directly present in the encoder output

That is not true. transformer_encoder is called on already positionaly encoder tensor.
The original paper does propably differently because they use RNNs. But this should work.

Reply:
As a side-note, I am referencing https://github.com/pytorch/pytorch/issues/24826 . Before, I didn't know that the pytorch transformer implementation didn't have PositionalEncoding implemented inside the abstraction.